pkg/operator/status.go

```diff
@@ -289,9 +299,172 @@ func (optr *Operator) syncUpgradeableStatus() error {
+		}
+	}
 	return optr.updateStatus(co, coStatus)
 }
 
+// isKubeletSkewSupported checks the version skew of kube-apiserver and node kubelet version.
+// Returns the skew status. version skew > 2 is not supported.
+func (optr *Operator) isKubeletSkewSupported(pools []*v1.MachineConfigPool) (skewStatus string, coStatus configv1.ClusterOperatorStatusCondition, err error) {
+	coStatus = configv1.ClusterOperatorStatusCondition{}
+	kubeAPIServerStatus, err := optr.configClient.ConfigV1().ClusterOperators().Get(context.TODO(), "kube-apiserver", metav1.GetOptions{})
+	if err != nil {
+		coStatus.Reason = skewUnchecked
+		coStatus.Message = fmt.Sprintf("An error occurred when checking kubelet version skew: %v", err)
+		return skewUnchecked, coStatus, err
+	}
+	// looks like
+	//   - name: kube-apiserver
+	//     version: 1.21.0-rc.0
+	kubeAPIServerVersion := ""
+	for _, version := range kubeAPIServerStatus.Status.Versions {
+		if version.Name != "kube-apiserver" {
+			continue
+		}
+		kubeAPIServerVersion = version.Version
+		break
+	}
+	if kubeAPIServerVersion == "" {
+		err = fmt.Errorf("kube-apiserver does not yet have a version")
+		coStatus.Reason = skewUnchecked
+		coStatus.Message = fmt.Sprintf("An error occurred when checking kubelet version skew: %v", err.Error())
+		return skewUnchecked, coStatus, err
+	}
+	kubeAPIServerMinorVersion, err := getMinorKubeletVersion(kubeAPIServerVersion)
+	if err != nil {
+		coStatus.Reason = skewUnchecked
+		coStatus.Message = fmt.Sprintf("An error occurred when checking kubelet version skew: %v", err)
+		return skewUnchecked, coStatus, err
+	}
+	var (
+		lastError      error
+		kubeletVersion string
+	)
+	nodes, err := optr.GetAllManagedNodes(pools)
+	if err != nil {
+		err = fmt.Errorf("getting all managed nodes failed: %v", err)
+		coStatus.Reason = skewUnchecked
+		coStatus.Message = fmt.Sprintf("An error occurred when getting all the managed nodes: %v", err.Error())
+	}
+	for _, node := range nodes {
+		// looks like kubeletVersion: v1.21.0-rc.0+6143dea
+		kubeletVersion = node.Status.NodeInfo.KubeletVersion
+		if kubeletVersion == "" {
+			continue
+		}
+		nodeMinorVersion, err := getMinorKubeletVersion(kubeletVersion)
+		if err != nil {
+			lastError = err
+			continue
+		}
+		if nodeMinorVersion+2 < kubeAPIServerMinorVersion {
+			coStatus.Reason = skewUnsupported
+			coStatus.Message = fmt.Sprintf("One or more nodes have an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet version of at least %v.", getMinimalSkewSupportNodeVersion(kubeAPIServerVersion))
+			return skewUnsupported, coStatus, nil
+		}
+		if nodeMinorVersion+2 == kubeAPIServerMinorVersion {
+			coStatus.Reason = skewPresent
+			coStatus.Message = fmt.Sprintf("Current kubelet version %v will not be supported by newer kube-apiserver. Please upgrade the the kubelet first if plan to upgrade the kube-apiserver", kubeletVersion)
+			return skewPresent, coStatus, nil
+		}
```

Comment:
Looks like for `nodeMinorVersion == kubeAPIServerMinorVersion`, this function will return skewSupported which couldn't be confusing sine there is no skew at all. Shouldn't we return `noSkew` for this case?

Reply:
renamed `skewSupported` to `noSkew`.

Reply:
Are you suggesting `nodeMinorVersion == kubeAPIServerMinorVersion` shouldbe `noSkew` and `kubeAPIServerMinorVersion - nodeMinorVersion == 1` shoule be `skewSupported`?

For `coStatus`, `kubeAPIServerMinorVersion - nodeMinorVersion < 2` is out of the scoop of this PR https://github.com/openshift/machine-config-operator/pull/2658#discussion_r675082603, I don't see it is necessary for `isKubeletSkewSupported` to descrimintate it at this moment.

Reply:
@sinnykumari I think you misread the line perhaps? there is a +2 which would be skewPresent.

Reply:
What I meant was to add a new field `noSkew` to handle condition `nodeMinorVersion == kubeAPIServerMinorVersion`. If you replace it, that would mean for `nodeMinorVersion+1 == kubeAPIServerMinorVersion` as well it will return noSkew, although there is skewPresent. I think @rphillips  can tell better what is best here. Because in the end how these status is being consumed and message we want to give customer matters.

Reply:
I revert the rename change, keep using `skewSupported` for `kubeAPIServerMinorVersion - nodeMinorVersion < 2`  until a consensus is reached.

Reply:
@sinnykumari is this now ok?

Reply:
yeah, based on https://github.com/openshift/machine-config-operator/pull/2658#discussion_r675082603 it seems handling n<2 is not part of current enhancement. So, it is upto node team how they get handled in 4.9.